src/dune_rules/exe_rules.ml

```diff
@@ -134,45 +142,45 @@ let executables_rules ~sctx ~dir ~expander ~dir_contents ~scope ~compile_info
       ~modules ~flags ~requires_link ~requires_compile ~preprocessing:pp
       ~js_of_ocaml ~opaque:Inherit_from_settings ~dynlink ~package:exes.package
   in
-  let o_files =
-    if not (Executables.has_foreign exes) then
-      []
-    else
-      let what =
-        if List.is_empty exes.buildable.Dune_file.Buildable.foreign_stubs then
-          "archives"
-        else
-          "stubs"
-      in
-      if List.mem Exe.Linkage.byte ~set:linkages then
-        User_error.raise ~loc:exes.buildable.loc
-          [ Pp.textf "Pure bytecode executables cannot contain foreign %s." what
-          ]
-          ~hints:
-            [ Pp.text
-                "If you only need to build a native executable use \"(modes \
-                 exe)\"."
-            ];
-      let foreign_sources =
-        let foreign_sources = Dir_contents.foreign_sources dir_contents in
-        Foreign_sources.for_exes foreign_sources ~first_exe
-      in
-      let o_files =
-        Foreign_rules.build_o_files ~sctx ~dir ~expander
-          ~requires:requires_compile ~dir_contents ~foreign_sources
-        |> List.map ~f:Path.build
-      in
-      Check_rules.add_files sctx ~dir o_files;
-      o_files
-  in
   let requires_compile = Compilation_context.requires_compile cctx in
   let preprocess =
     Preprocess.Per_module.with_instrumentation exes.buildable.preprocess
       ~instrumentation_backend:
         (Lib.DB.instrumentation_backend (Scope.libs scope))
   in
-  Exe.build_and_link_many cctx ~programs ~linkages ~link_args ~o_files
-    ~promote:exes.promote ~embed_in_plugin_libraries;
+  let () =
+    (* TODO: Currently [exe_rules] differ from [lib_rules] in some aspects and
+       the reason is unclear. For example, instead of building an archive for
+       foreign stubs, we link the corresponding object files directly. It would
+       be nice to make the code more uniform. *)
```

Comment:
Factorising the code seems good, but going through an intermediate .a file for executables feels a bit unnecessary. It would be the same as going through a `.cmxa` for building an executable, which feels a bit heavy.

Reply:
Note that this comment was just moved to a new spot. It already existed before this PR. I ended up fixing it after your comment however.